test/jasmine/tests/layout_images_test.js

```diff
@@ -290,6 +294,31 @@ describe('Layout images', function() {
 
         afterEach(destroyGraphDiv);
 
+        it('should only create canvas if url image', function(done) {
+            var originalCreateElement = document.createElement;
+            var newCanvasElement;
+            spyOn(document, 'createElement').and.callFake(function(elementType) {
+                var element = originalCreateElement.call(document, elementType);
+                if(elementType === 'canvas') {
+                    newCanvasElement = element;
+                    spyOn(newCanvasElement, 'toDataURL');
+                }
+                return element;
+            });
+
+            Plotly.relayout(gd, 'images[0].source', dataUriImage).then(function() {
```

Comment:
Non-blocking but something like:

```js
        it('should only create canvas if url image', function(done) {
            var originalCreateElement = document.createElement;
            var newCanvasElement;
            spyOn(document, 'createElement').and.callFake(function(elementType) {
                var element = originalCreateElement.call(document, elementType);
                if(elementType === 'canvas') {
                    newCanvasElement = element;
                    spyOn(newCanvasElement, 'toDataURL');
                }
                return element;
            });

            Plotly.relayout(gd, 'images[0].source', dataUriImage)
            .then(function() {
                expect(newCanvasElement).toBeUndefined();
            })
            .then(function() { return Plotly.relayout(gd, 'images[0].source', pythonLogo); })
            .then(function() {
                expect(newCanvasElement).toBeDefined();
                expect(newCanvasElement.toDataURL).toHaveBeenCalledTimes(1);
            })
            .catch(failTest)
            .then(done)
        });
```

should work.

Reply:
The trouble there is that the first `expect(newCanvasElement).toBeUndefined();` will also pass for a url image, because the `onload` callback won't necessarily have completed yet. At least that's what I concluded was happening when I tried this.

Reply:
> The trouble there is that the first `expect(newCanvasElement).toBeUndefined();` will also pass for a url image, because the `onload` callback won't necessarily have completed yet.

Hmm, that shouldn't happen, the `Plotly.relayout` promise shouldn't resolve before the `onload` callback is executed.

Reply:
Ok, I'll double check...

Reply:
Yeah, you're right. Not sure what I was trying before.  I updated the test to use this structure.